python/ray/tune/pbt.py

```diff
@@ -0,0 +1,174 @@
+                    config = config_dictionary,
+                    local_dir= best_trial.local_dir,
+                    experiment_tag= best_trial.experiment_tag + mutate_string,
+                    resources = best_trial.resources,
+                    stopping_criterion= best_trial.stopping_criterion,
+                    checkpoint_freq= best_trial.checkpoint_freq,
+                    restore_path= checkpoint,
+                    upload_dir= best_trial.upload_dir
+                )
+                
+                #mutate the hyperparamters the new trial, last index of config is always the "env"
+                #so exclude that, temporarily we "perturb" each hyperparameter by a factor 1.2 or 0.8 and then round it to the nearest
+                #even integer
+                print(new_trial.config)
+                new_trial.experiment_tag = trial.experiment_tag + "-mutated@" + str(time)
+                #add the new trial to the trialrunner, and then stop the
+                #  trial it replaced
+                trial_runner._launch_trial(new_trial)
+                trial_runner.add_trial(new_trial)
+                self._results[new_trial] = self._results[trial]
+                self._time_tracker[new_trial] = self._time_tracker[trial]
+                self._results.pop(trial)
+                self._time_tracker.pop(trial)
+                best_trial.stop() if best_trial_was_stopped else print("best trial continues")
+                return TrialScheduler.STOP
+                
+
+
+        else:
+            print("not ready")
+        
+        return TrialScheduler.CONTINUE
+
+    def on_trial_complete(self, trial_runner, trial, result):
+        self._results[trial].append(result)
+        self._completed_trials.add(trial)
+        print("finished a trial")
+        min_time = 0
+        best_trial = None
+        for trial in self._completed_trials:
+            last_result = self._results[trial][-1]
+            if getattr(last_result, self._time_attr) < min_time or min_time == 0:
+                min_time = getattr(last_result, self._time_attr)
+                best_trial = trial
+        print("The Best Trial is currently {0} finishing in {1} iterations, with the hyperparameters of {2}".format(best_trial, min_time, best_trial.config))
+
+    def _default_ready(self, result, trial, time, time_diff = 4):
+        #function checks if trial has been added to the time tracker, if not adds
+        #it, else checks if an appropriate amount of time has passed since the last mutation
+        if trial not in self._time_tracker:
+            print("added trial to time tracker")
+            self._time_tracker[trial] = (time)
+        else:
+            time_since_last = time - self._time_tracker[trial]
+            if time_since_last >= time_diff:
+                self._time_tracker[trial] = time
+                return True
+        return False
+    
+    def _get_best_trial(self, result, time):
```

Comment:
From what I understand, the algorithm here does a relative comparison to find the best trial, and uses that best trial to mutate the current weights. In example (in my reading of the code): given`trial`, `_get_best_trial` returns a best trial conditioned on the current iteration `trial.current_iteration` - this given trial might not actually be good in real time though. 

What are your thoughts?

Reply:
My thought process on the _get_best_trial function was that it would be unfair to compare at a further along iteration since it is not compared to all the other hyperparameters/weights on the same level field. That may not be entirely correct though.